base/reduce.jl

```diff
@@ -231,6 +231,13 @@ guaranteed left or right associativity and invocation of `f` for every value.
 """
 mapreduce(f, op, v0, itr) = mapfoldl(f, op, v0, itr)
 
+mapreduce(mapper, reducer, args::Tuple{}) = _empty_reduce_error()
+mapreduce(mapper, reducer, args::Tuple) =
+    mapreduce(mapper, reducer, mapper(first(args)), tail(args))
+mapreduce(mapper, reducer, default, args::Tuple{}) = default
+mapreduce(mapper, reducer, default, args::Tuple) =
+    reducer(mapper(first(args)), mapreduce(mapper, reducer, default, tail(args)))
```

Comment:
To match the order used by other methods of `mapreduce`, this should be `mapreduce(mapper, reducer, reducer(default, first(args)), tail(args))`.

Reply:
That makes sense. In that case wouldn't it make sense to make these methods of `foldl`?

Reply:
Ah yes, the order is undefined for `mapreduce`, so this is technically ok. But I find this particular order pretty strange, e.g.

```
julia> mapreduce(identity, tuple, (1,2,3,4))
(2, (3, (4, 1)))
```

Maybe `reducer(default, mapreduce(mapper, reducer, mapper(first(args)), tail(args)))`, which gives `foldr` order.

Another issue is that fully-unrolled implementations usually don't work well for big tuples. For example see the implementation of tuple `map` in tuple.jl. Unfortunately we don't have a good general solution to this yet.

Reply:
So like this?

```julia
mapfoldr(mapper, reducer, args::Tuple{}) = _empty_reduce_error()
mapfoldr(mapper, reducer, args::Tuple) =
    mapfoldr(mapper, reducer, mapper(first(args)), tail(args))
mapfoldr(mapper, reducer, default, args::Tuple{}) = default
mapfoldr(mapper, reducer, default, args::Tuple) =
    reducer(default, mapfoldr(mapper, reducer, mapper(first(args)), tail(args)))

mapreduce(mapper, reducer, args::Tuple) = mapfoldr(mapper, reducer, args)
mapreduce(mapper, reducer, default, args::Tuple) = 
    mapfoldr(mapper, reducer, default, args)
```

Reply:
> Unfortunately we don't have a good general solution to this yet.

Actually, I suspect we do now: take out the `@inline`. In theory the new cost model should do the right thing (and if not, we should try to fix the cost model).

Reply:
Ok, figured out what was going wrong with hashing:

I had to change

`hash(r::MersenneTwister, h::UInt) = foldr(hash, h, (r.seed, r.state, r.vals, r.idx))`

to

`reduce(hash, (r.seed, r.state, r.vals, r.idx, h))`

because this slightly changes foldr semantics such that the default argument will be the first in the tuple and not the last.